Introduce map_get guard-safe function

michalmuskala opened the pull request:
## Rationale

Today all compound data types except for maps can be deconstructed in guards.
For tuples we have `element/2` and for lists `hd/1` and `tl/1`. Maps are
completely opaque to guards. This means matching on maps can't be
abstracted into macros, which is often done with repetitive guards. It
also means that maps have to be always selected whole from ETS tables,
even when only one field would be enough, which creates a potential
efficiency issue.

This PR introduces an `erlang:map_get/2` guard-safe function that allows
extracting a map field in guard. An alternative to this function would be
to introduce the syntax for extracting a value from a map that was planned
in the original EEP: `Map#{Key}`.

Even outside of guards, since this function is a guard-BIF it is more
efficient than using `maps:get/2` (since it does not need to set up the
stack), and more convenient from pattern matching on the map (compare:
`#{key := Value} = Map, Value` to `map_get(key, Map)`).

## Performance considerations

A common concern against adding this function is the notion that "guards
have to be fast" and ideally execute in constant time. While there are
some counterexamples (`length/1`), what is more important is the fact
that adding those functions does not change in any way the time
complexity of pattern matching - it's already possible to match on map
fields today directly in patterns - adding this ability to guards will
niether slow down or speed up the execution, it will only make certain
programs more convenient to write.

This first version is very naive and does not perform any optimizations.

## TODO

* [ ] compile to `get_map_elements` when safe - this will also compute the
  hash value at compile-time if applicable
* [x] allow use in match specs
* [ ] use knowledge that the second argument is a map in strength
  reduction compiler passes
* [x] documentation

bjorng commented:
The idea seems reasonable to me and the implementations looks good.

OTP Technical Board will discuss this pull request on Thursday this week.

A minor thing regarding the implementation: If you include `maps_get/2` in `erl_bifs:is_pure/3`, `sys_core_fold` will automatically do some basic optimizations.

bjorng commented on erts/emulator/beam/bif.c:
```diff
@@ -5275,3 +5275,48 @@ BIF_RETTYPE dt_restore_tag_1(BIF_ALIST_1)
     BIF_RET(am_true);
 }
+
+BIF_RETTYPE map_get_2(BIF_ALIST_2) {
```
Why duplicate so much code?

I suggest that you make a copy of `maps_get_2()` in `erl_map.c` and rename it to `map_get_2()`.

bjorng commented:
OTP Technical Board has approved `map_get/2`.

We will need the following things updated as soon as possible in order for us to test it in our daily builds before releasing the first release candidate:

* Elimination of the code duplication as in my previous comment.
* Updated `erl_bifs:is_pure/1`.
* Documentation.

Of the other things in your TODO list, being able to use `map_get/2` in a match spec is the most important one. It would be nice, but not necessary, to include that too in OTP 21.

bjorng commented:
I am not an expert on match specs myself, but it seems that main test suite for match specs is `match_spec_SUITE` in the emulator test suite. There already is a test case called `maps/1` in that module that you could extend.

`match_spec_SUITE` does most of its test by calling `ets:match_spec_test/3`, so it could also make sense to add some simple test of maps in match specs to the tests of `ets:select()` in `ets_SUITE` in STDLIB.

michalmuskala commented:
Ok, I've pushed commits implementing the things you asked about

* Duplicate less code in the implementation itself,
* Basic compiler optimisation by marking `map_get` as pure
* Document the guard
* Document use in match specs (for some reason `map_size` wasn't documented there earlier, so I added it as well)
* Add tests for use of `map_get` and `map_size` in match specs

bjorng commented:
@michalmuskala The documentation doesn't build.

While you are at it, please squash the commits. Unless the daily builds show any problems, I want to merge this pull request early next week.

bjorng commented:
@michalmuskala After updating the primary bootstrap, there is a compiler warning for `erl_types.erl`, and a subsequent build failure because of warnings as errors. The following patch fixes the problem:

```
diff --git a/lib/hipe/cerl/erl_types.erl b/lib/hipe/cerl/erl_types.erl
index 8a609ef..a91da97 100644
--- a/lib/hipe/cerl/erl_types.erl
+++ b/lib/hipe/cerl/erl_types.erl
@@ -217,7 +217,7 @@
        ]).
 
 %%-define(DO_ERL_TYPES_TEST, true).
--compile({no_auto_import,[min/2,max/2]}).
+-compile({no_auto_import,[min/2,max/2,map_get/2]}).
 
 -ifdef(DO_ERL_TYPES_TEST).
 -export([test/0]).

```

bjorng commented:
@michalmuskala One more thing (hopefully final): I think that the implementation of `maps:get/2` should be moved from `bif.c` to `erl_map.c`. That will keep the map functionality in one source file (except for the GC version of `map_size/1`, which is defined in `erl_bif_guard.c`). The non-GCing version of `map_size/1` is defined in `erl_map.c`. Another reason is that the C compiler could do some inlining.

michalmuskala commented:
I did the fix in `erl_types.erl` and checked that the docs compile.

I moved the implementation of `erlang:map_get/2` to `erl_map.c`, extracted an inlined `common_maps_get` function and made both BIFs (`erlang:map_get/2` and `maps:get/2`) call that.

bjorng commented:
@michalmuskala Looks good.

Only one more thing - the TODO in the commit message is not updated.

bjorng commented:
Thanks! Added to our daily builds.

garazdawi commented on erts/emulator/beam/erl_map.c:
```diff
@@ -202,25 +203,33 @@ BIF_RETTYPE maps_find_2(BIF_ALIST_2) {
 
-BIF_RETTYPE maps_get_2(BIF_ALIST_2) {
-    if (is_map(BIF_ARG_2)) {
+BIF_RETTYPE ERTS_INLINE common_maps_get(Process *p, Eterm key, Eterm map) {
```
This function should be static, or put the entire implementation into maps_get_2 and make map_get_2 call it.

michalmuskala replied:
I tried to make one call the other but I couldn't figure out all the arguments to satisfy the BIF_ALIST_2 declaration.

garazdawi replied:
Use the BIF_CALL_ARGS macro. i.e. something like:

```
BIF_RETTYPE map_get_2(BIF_ALIST_2) {
  BIF_RET(maps_get_2(BIF_CALL_ARGS));
}
```

michalmuskala commented:
@garazdawi updated

garazdawi commented:
Failing testcases:
tools/emacs_SUITE - bif_highlight
emulator/bif_SUITE - spec + shadow_comments

bjorng commented:
@michalmuskala @garazdawi The `emulator/bif_SUITE` test cases fail because I didn't update `erlang.beam` when adding the branch to our daily builds. (Don't update `erlang.beam`; I will take care of that.)

I just noted that the indentation for the added in lines in `erl_bif_types` is off. Please fix that.

michalmuskala commented:
I've fixed the bug in `tools/emacs_SUITE` and the indentation in `erl_bif_types.erl`.
